File: SpellingBee.java
// import java.applet.*;
// import java.awt.*;
// This program solves the NYT's Spelling Bee Puzzle in a text-based UI.
// I might make it into a browser plugin.
import java.util.*;
import java.io.*;

public class SpellingBee {
    private static Scanner console;
    private Map<String, Set<Character>> wordMap;
    public int MIN_WORD_LENGTH = 4;
    public int MAX_ALLOWED_CHARACTERS = 7;

    public SpellingBee(Set<String> dictionary) throws FileNotFoundException {
        wordMap = new HashMap<String, Set<Character>>();
        console = new Scanner(System.in);
        for (String word : dictionary) {
            if (word.length() >= MIN_WORD_LENGTH) {
                Set<Character> set = new HashSet<Character>();
                for (int i = 0; i < word.length(); i++) {
                    set.add(word.charAt(i));
                }
                if (set.size() <= MAX_ALLOWED_CHARACTERS) {
                    wordMap.put(word, set);
                }
            }
        }
    }

    public Set<String> wordFinder(String letters, char middle) {
        letters = letters.toLowerCase();
        Set<Character> charSet = new HashSet<>();
        for (int i = 0; i < letters.length(); i++) {
            charSet.add(letters.charAt(i));
        }
        Set<String> result = new TreeSet<String>();
        for (String word : wordMap.keySet()) {
            Set<Character> set = wordMap.get(word);
            if (set.contains(middle) && charSet.containsAll(set)) {
                result.add(word);
            }
        }
        return result;
    }

    public boolean yesTo(String prompt) {
        System.out.print(prompt + " (y/n)? ");
        console = new Scanner(System.in);
        String response = console.nextLine().trim().toLowerCase();
        while (!response.equals("y") && !response.equals("n")) {
            System.out.println("Only type \"y\" or \"n\" (case-insensitive).");
            System.out.print(prompt + " (y/n)? ");
            response = console.nextLine().trim().toLowerCase();
        }
        return response.equals("y");
    }

    public static boolean hasDuplicateLetters(String s) {
        Set<Character> charSet = new HashSet<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (charSet.contains(c)) {
                return true;
            }
            charSet.add(c);
        }
        return false;
    } 

    public static void interact(SpellingBee bee) {
        console = new Scanner(System.in);
        System.out.print("Letters? ");
        String response = console.next();
        while (response.length() != 7 || hasDuplicateLetters(response)) {
            System.out.println("Invalid input (n != 7 or duplicate letters).");
            System.out.print("Letters? ");
            response = console.next();
        }
        System.out.print("Middle character? ");
        String ch = console.next();
        while (ch.length() != 1) {
            System.out.println("Invalid input.");
            System.out.print("Middle character? ");
            ch = console.next();
        }
        System.out.println(bee.wordFinder(response, ch.charAt(0)));
        if (bee.yesTo("Do you want to go again?")) {
            interact(bee);
        }
    }

    public static void main(String[] args) throws FileNotFoundException {
        File file = new File("dict.txt");
        Set<String> dictionarySet = new HashSet<String>();
        Scanner scanner = new Scanner(file);
        while (scanner.hasNextLine()) {
            dictionarySet.add(scanner.nextLine());
        }
        scanner.close();
        interact(new SpellingBee(dictionarySet));
    }
}

File: SpellingBee.py
import sys
from collections import defaultdict

class SpellingBee:
    MIN_WORD_LENGTH = 4
    MAX_ALLOWED_CHARACTERS = 7

    def __init__(self, dictionary):
        self.word_map = defaultdict(set)
        for word in dictionary:
            if len(word) >= self.MIN_WORD_LENGTH:
                char_set = set(word)
                if len(char_set) <= self.MAX_ALLOWED_CHARACTERS:
                    self.word_map[word] = char_set

    def word_finder(self, letters, middle):
        letters = letters.lower()
        char_set = set(letters)
        result = set()
        for word, word_set in self.word_map.items():
            if middle in word_set and char_set.issuperset(word_set):
                result.add(word)
        return result

    def pangram_finder(self, letters):
        letters = letters.lower()
        char_set = set(letters)
        result = set()
        for word, word_set in self.word_map.items():
            if char_set == word_set:
                result.add(word)
        return result

    def yes_to(self, prompt):
        response = input(prompt + " (y/n)? ").strip().lower()
        while response not in {"y", "n"}:
            print("Only type \"y\" or \"n\" (case-insensitive).")
            response = input(prompt + " (y/n)? ").strip().lower()
        return response == "y"

    @staticmethod
    def has_duplicate_letters(s):
        char_set = set()
        for c in s:
            if c in char_set:
                return True
            char_set.add(c)
        return False

    def interact(self):
        response = input("Letters? ")
        while len(response) != 7 or self.has_duplicate_letters(response):
            print("Invalid input (n != 7 or duplicate letters).")
            response = input("Letters? ")
        ch = input("Middle character? ")
        while len(ch) != 1:
            print("Invalid input.")
            ch = input("Middle character? ")
        if self.yes_to("Pangrams only?"):
            print(self.pangram_finder(response))
        else:
            print(self.word_finder(response, ch[0]))
        if self.yes_to("Do you want to go again?"):
            self.interact()


if __name__ == "__main__":
    # Load the dictionary and create a SpellingBee instance
    dictionary = set()
    with open("dict.txt", "r") as f:
        for line in f:
            dictionary.add(line.strip())
    spelling_bee = SpellingBee(dictionary)
    spelling_bee.interact()
